scripts/py_matter_yamltests/matter_yamltests/parser.py

```diff
@@ -490,42 +536,86 @@ def busy_wait_ms(self):
     def wait_for(self):
         return self._test.wait_for
 
-    def post_process_response(self, response: dict):
+    @property
+    def event_number(self):
+        return self._test.event_number
+
+    def post_process_response(self, received_responses):
         result = PostProcessResponseResult()
 
+        # A list of responses is what is expected, but for legacy, if the response
+        # does not comes up as a list, it is converted here.
+        # TODO It should be removed once all decoders returns a list.
+        if not isinstance(received_responses, list):
+            received_responses = [received_responses]
+
         if self.wait_for is not None:
-            self._response_cluster_wait_validation(response, result)
+            self._response_cluster_wait_validation(received_responses, result)
             return result
 
-        if self._skip_post_processing(response, result):
+        if self._skip_post_processing(received_responses, result):
             return result
 
-        self._response_error_validation(response, result)
-        if self.response:
-            self._response_cluster_error_validation(response, result)
-            self._response_values_validation(response, result)
-            self._response_constraints_validation(response, result)
-            self._maybe_save_as(response, result)
+        check_type = PostProcessCheckType.RESPONSE_VALIDATION
+        error_failure_wrong_response_number = f'The test expects {len(self.responses)} responses but got {len(received_responses)} responses.'
+
+        received_responses_copy = copy.deepcopy(received_responses)
+        for expected_response in self.responses:
+            if len(received_responses_copy) == 0:
+                result.error(check_type, error_failure_wrong_response_number)
+                return result
+            received_response = received_responses_copy.pop(0)
+            self._response_error_validation(
+                expected_response, received_response, result)
+            self._response_cluster_error_validation(
+                expected_response, received_response, result)
+            self._response_values_validation(
+                expected_response, received_response, result)
+            self._response_constraints_validation(
+                expected_response, received_response, result)
+            self._maybe_save_as(expected_response, received_response, result)
+
+        # An empty response array in a test step (responses: []) implies that the test step does expect a response
+        # but without any associated value.
+        if self.responses == [] and received_responses_copy == [{}]:
+            # if the received responses is a simple success ([{}]), that is valid.
+            return result
+        # This is different from the case where no response is specified at all, which implies that the step expect
+        # a success with any associatied value(s).
```

Comment:
This comment confuses me.  It's not describing the case guarded by `elif` that follows it (right?), but it's indented as if it were.  What case is this describing and how is it represented in this code?

Reply:
Actually it does describe the case guarded by `elif`. It may be confusing because the `[{}]` has been converted to `[{'values': [{}]}]` by `TestStepWithPlaceholders::_convert_single_value_to_values`